### PR TITLE
Validate lint docs separately.

### PR DESCRIPTION
**Pull request by ehuss**
This addresses some concerns raised in https://github.com/rust-lang/rust/pull/76549#issuecomment-727638552 about errors with the lint docs being confusing and cumbersome. Errors from validating the lint documentation were being generated during `x.py doc` (and `x.py dist`), since extraction and validation are being done in a single step. This changes it so that extraction and validation are separated, so that `x.py doc` will not error if there is a validation problem, and tests are moved to `x.py test src/tools/lint-docs`.

This includes the following changes:

* Separate validation to `x.py test`.
* Added some more documentation on how to more easily modify and test the docs.
* Added more help to the error messages to hopefully provide more information on how to fix things.

The first commit just moves the code around, so you may consider looking at the other commits for a smaller diff.

### CODE REVIEW

**Comment by ehuss** [comment]
cc @petrochenkov FYI

**Review by Mark-Simulacrum — COMMENTED** [review]
r=me with nits fixed (feel free to ignore strip_prefix though if it doesn't seem like net improvement after modifying).

File: src/tools/lint-docs/src/lib.rs [hunk 1]
```diff
@@ -45,382 +83,397 @@ impl Level {
...
+                }
+            };
+            // Read the lint.
+            let mut doc_lines = Vec::new();
+            let (doc, name) = loop {
+                match lines.next() {
+                    Some((lineno, line)) => {
+                        let line = line.trim();
+                        if line.starts_with("/// ") {
+                            doc_lines.push(line.trim()[4..].to_string());
```
[the remark this answers is not in the document]
Comment (ehuss) [reply src/tools/lint-docs/src/lib.rs]: Yea, I think the strip_prefix is better. I'm not sure why, but I keep forgetting about it.

**Pull request merged**